unyt/dask_array.py

```diff
@@ -0,0 +1,620 @@
+        # call to the ufunc with the unyt_inputs will enforce the unyt rules
+        # (e.g., addition must have same dimensions).
+        ui_0, ui_1 = unyt_inputs[0], unyt_inputs[1]
+        if _check_unyt_inputs(ui_0, ui_1):
+
+            # convert to the unit with the larger base
+            input = list(input)
+            if ui_0.units.base_value < ui_1.units.base_value:
+                input[0] = input[0].to(ui_1.units)
+            else:
+                input[1] = input[1].to(ui_0.units)
+            unyt_inputs = _extract_unyt(input)
+
+    return input, unyt_inputs
+
+
+def _prep_ufunc(ufunc, *input, extract_dask=False, **kwargs):
+    # this function:
+    # (1) sanitizes inputs for calls to __array_func__, __array__ and _elementwise
+    # (2) applies the function to the hidden unyt quantities
+    # (3) (optional) makes inputs extra clean: converts unyt_dask_array args to
+    #     plain dask array objects
+
+    # apply the operation to the hidden unyt_quantities
+    input, unyt_inputs = _sanitize_unit_args(*input)
+    unyt_result = ufunc(*unyt_inputs, **kwargs)
+
+    if extract_dask:
+        input = _extract_dask(input)
+
+    input = _extract_unyt_val(input)
+    return input, unyt_result
+
+
+def _post_ufunc(dask_superfunk, unyt_result):
+    # a decorator to attach hidden unyt quantity to result of a ufunc, array or
+    # elemwise calculation
+    def wrapper(*args, **kwargs):
+        dask_result = dask_superfunk(*args, **kwargs)
+        if hasattr(unyt_result, "units"):
+            return _create_with_quantity(dask_result, unyt_result)
+        return dask_result
+
+    return wrapper
+
+
+def _special_dec(the_func):
+    # decorator for special operations like __mul__ , __truediv__, which will
+    # bypass __getattribute__
+    def wrapper(self, *args, **kwargs):
+        funcname = the_func.__name__
+        ufunc = getattr(ua.unyt_quantity, funcname)
+        ufunc_args = (self,) + args
+        args, unyt_result = _prep_ufunc(ufunc, *ufunc_args, extract_dask=True, **kwargs)
+
+        # remove the first arg, cause we need to supply self as first arg
+        _ = args.pop(0)
+        daskresult = the_func(self, *args, **kwargs)
+
+        if hasattr(unyt_result, "units"):
+            return _create_with_quantity(daskresult, unyt_result)
+        return daskresult  # __lt__, __le__, etc. will hit this
+
+    return wrapper
+
+
+class unyt_dask_array(_dask_Array):
+    """
+    a dask.array.core.Array subclass that tracks units. Easiest to use the
+    unyt_from_dask helper function to generate new instances.
+
+    Parameters
+    ----------
+
+    All parameters are those for dask.array.core.Array
+
+    """
+
+    def __new__(
+        clss,
+        dask_graph,
+        name,
+        chunks,
+        dtype=None,
+        meta=None,
+        shape=None,
+        units=None,
+        registry=None,
+        bypass_validation=False,
+        input_units=None,
+        unyt_name=None,
```

Comment:
my interpretation of the docs is that this class isn't intended as publicly exposed, but it feels safer to treat it as if it was.
```suggestion
        *,
        dtype=None,
        meta=None,
        shape=None,
        units=None,
        registry=None,
        bypass_validation=False,
        input_units=None,
        unyt_name=None,
```
Also, it seems that a lot of these arguments are not re-exposed via `unyt_from_dask`, which I understand is the actual public interface. Should they be removed or exposed ?

Reply:
will have to double check if there a reason not too, but looking now I do think all these kwargs should be in `unyt_from_dask` as well.